ygnmi/gnmi.go

```diff
@@ -330,25 +333,49 @@ const (
 )
 
 // populateSetRequest fills a SetResponse for a val and operation type.
-func populateSetRequest(req *gpb.SetRequest, path *gpb.Path, val interface{}, op setOperation, preferShadowPath bool) error {
+func populateSetRequest(req *gpb.SetRequest, path *gpb.Path, val interface{}, op setOperation, preferShadowPath bool, opts ...Option) error {
 	if req == nil {
 		return fmt.Errorf("cannot populate a nil SetRequest")
 	}
+	opt := resolveOpts(opts)
 
 	switch op {
 	case deletePath:
 		req.Delete = append(req.Delete, path)
 	case replacePath, updatePath:
-		// Since the GoStructs are generated using preferOperationalState, we
-		// need to turn on preferShadowPath to prefer marshalling config paths.
-		js, err := ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
-		if err != nil {
-			return fmt.Errorf("could not encode value into JSON format: %w", err)
+		var typedVal *gpb.TypedValue
+		var err error
+		if opt.preferProto {
+			typedVal, err = ygot.EncodeTypedValue(val, gpb.Encoding_JSON_IETF, &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+		} else {
+			typedVal = &gpb.TypedValue{Value: &gpb.TypedValue_JsonIetfVal{}}
+			// Since the GoStructs are generated using preferOperationalState, we
+			// need to turn on preferShadowPath to prefer marshalling config paths.
+			typedVal.Value.(*gpb.TypedValue_JsonIetfVal).JsonIetfVal, err = ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+		}
+
+		if err != nil && opt.setFallback && path.Origin != "openconfig" {
```

Comment:
Check origin is not ""

Reply:
resolvePath always sets the origin to "openconfig" if it empty https://github.com/openconfig/ygnmi/blob/main/ygnmi/gnmi.go#L410

Reply:
The current unit test seems to interpret "" as non-openconfig.

Reply:
fixed and added test